In [1]:
import pandas as pd
from sklearn.cluster import DBSCAN
from sklearn.preprocessing import StandardScaler

# קריאת הקובץ
data = pd.read_csv(r"C:\Users\guymk\Downloads\data13 (1).csv")

# בחירת עמודות רלוונטיות לקלסטרינג: HitTime, Arc, ShotDistance (מהעמודות "Raw Data" והעמודות הרלוונטיות)
data_filtered = data[['Raw Data', 'Unnamed: 5', 'Unnamed: 28']].copy()

# שינוי שמות העמודות למשהו אינטואיטיבי יותר
data_filtered.columns = ['HitTime', 'Arc', 'ShotDistance']

# מחיקת שורות עם ערכים חסרים בעמודות אלו
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# המרת 'HitTime' לערכים מספריים (בהנחה שזהו timestamp או נתונים מספריים)
data_filtered['HitTime'] = pd.to_numeric(data_filtered['HitTime'], errors='coerce')

# מחיקת ערכים חסרים
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# נרמול העמודות 'Arc' ו-'ShotDistance' לצורך DBSCAN
scaler = StandardScaler()
X_filtered_scaled = scaler.fit_transform(data_filtered[['HitTime', 'Arc', 'ShotDistance']])

# הרצת DBSCAN על הדאטה המתוקן
dbscan_filtered_model = DBSCAN(eps=0.5, min_samples=5)
clusters_filtered = dbscan_filtered_model.fit_predict(X_filtered_scaled)

# הוספת תוצאות הקלסטרינג לדאטה
data_filtered['Cluster'] = clusters_filtered

# הצגת התוצאות
data_filtered.head()


HitTime   Arc ShotDistance  Cluster
1  1.592350e+12  -101          113        0
2  1.592350e+12     4          162        0
3  1.592350e+12   -68           32        0
4  1.592350e+12   -63          115        0
5  1.592350e+12   -70          141        0

In [2]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import DBSCAN
import ace_tools as tools

# קריאת הדאטה
data = pd.read_csv(r"C:\Users\guymk\Downloads\data13 (1).csv")

# סינון והכנת הנתונים (התאמת עמודות)
data_filtered = data[['Raw Data', 'Unnamed: 5', 'Unnamed: 28']].copy()
data_filtered.columns = ['HitTime', 'Arc', 'ShotDistance']
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# המרת העמודה HitTime לערכים מספריים
data_filtered['HitTime'] = pd.to_numeric(data_filtered['HitTime'], errors='coerce')
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# נרמול הנתונים
scaler = StandardScaler()
X_filtered_scaled = scaler.fit_transform(data_filtered[['HitTime', 'Arc', 'ShotDistance']])

# הרצת DBSCAN עם כוונון הפרמטרים:
eps_value = 0.4  # אפשר לנסות גם ערכים של 0.3 או 0.5
min_samples_value = 5  # אפשר לנסות להגדיל את הערך בהתאם

dbscan_optimized = DBSCAN(eps=eps_value, min_samples=min_samples_value)
clusters_optimized = dbscan_optimized.fit_predict(X_filtered_scaled)

# הוספת הקלאסטרים לדאטה
data_filtered['Cluster_Optimized'] = clusters_optimized

# הצגת התוצאות עם הקלאסטרים החדשים
tools.display_dataframe_to_user(name="Optimized DBSCAN Clustering Results", dataframe=data_filtered)

# הצגת השורות הראשונות עם הקלאסטרים
print(data_filtered[['HitTime', 'Arc', 'ShotDistance', 'Cluster_Optimized']].head())


ModuleNotFoundError: No module named 'ace_tools'

In [3]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import DBSCAN

# קריאת הדאטה
data = pd.read_csv(r"C:\Users\guymk\Downloads\data13 (1).csv")  # עדכן את הנתיב בהתאם לצורך

# סינון והכנת הנתונים (התאמת עמודות)
data_filtered = data[['Raw Data', 'Unnamed: 5', 'Unnamed: 28']].copy()
data_filtered.columns = ['HitTime', 'Arc', 'ShotDistance']
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# המרת העמודה HitTime לערכים מספריים
data_filtered['HitTime'] = pd.to_numeric(data_filtered['HitTime'], errors='coerce')
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# נרמול הנתונים
scaler = StandardScaler()
X_filtered_scaled = scaler.fit_transform(data_filtered[['HitTime', 'Arc', 'ShotDistance']])

# הרצת DBSCAN עם כוונון הפרמטרים:
eps_value = 0.4  # אפשר לנסות גם ערכים של 0.3 או 0.5
min_samples_value = 5  # אפשר לנסות להגדיל את הערך בהתאם

dbscan_optimized = DBSCAN(eps=eps_value, min_samples=min_samples_value)
clusters_optimized = dbscan_optimized.fit_predict(X_filtered_scaled)

# הוספת הקלאסטרים לדאטה
data_filtered['Cluster_Optimized'] = clusters_optimized

# הצגת התוצאות עם הקלאסטרים החדשים
print(data_filtered[['HitTime', 'Arc', 'ShotDistance', 'Cluster_Optimized']].head())

# אם תרצה לראות יותר שורות מהדאטה
print(data_filtered[['HitTime', 'Arc', 'ShotDistance', 'Cluster_Optimized']].tail(10))


        HitTime   Arc ShotDistance  Cluster_Optimized
1  1.592350e+12  -101          113                  0
2  1.592350e+12     4          162                  0
3  1.592350e+12   -68           32                  0
4  1.592350e+12   -63          115                  0
5  1.592350e+12   -70          141                  0
            HitTime   Arc ShotDistance  Cluster_Optimized
13723  1.598240e+12  -163           56                  2
13724  1.598240e+12  -162           78                  2
13725  1.598240e+12  -131          104                  2
13726  1.598240e+12  -180          164                  2
13727  1.598240e+12  -180           32                  2
13728  1.598240e+12  -184           93                  2
13729  1.598240e+12  -165           92                  2
13730  1.598240e+12  -163           71                  2
13731  1.598240e+12  -212           26                  2
13732  1.598240e+12  -221           32                  2


## Silhouette Score

In [4]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import DBSCAN
from sklearn.metrics import silhouette_score

# קריאת הדאטה
data = pd.read_csv(r"C:\Users\guymk\Downloads\data13 (1).csv")  # עדכן את הנתיב בהתאם לצורך

# סינון והכנת הנתונים (התאמת עמודות)
data_filtered = data[['Raw Data', 'Unnamed: 5', 'Unnamed: 28']].copy()
data_filtered.columns = ['HitTime', 'Arc', 'ShotDistance']
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# המרת העמודה HitTime לערכים מספריים
data_filtered['HitTime'] = pd.to_numeric(data_filtered['HitTime'], errors='coerce')
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# נרמול הנתונים
scaler = StandardScaler()
X_filtered_scaled = scaler.fit_transform(data_filtered[['HitTime', 'Arc', 'ShotDistance']])

# טווח של ערכי eps ו-min_samples
eps_values = [0.3, 0.4, 0.5, 0.6]  # נסה מספר ערכים ל-eps
min_samples_values = [4, 5, 6]  # נסה מספר ערכים ל-min_samples

# איחסון התוצאות
best_eps = None
best_min_samples = None
best_sil_score = -1

# הרצת DBSCAN עבור כל שילוב של eps ו-min_samples
for eps_value in eps_values:
    for min_samples_value in min_samples_values:
        dbscan = DBSCAN(eps=eps_value, min_samples=min_samples_value)
        labels = dbscan.fit_predict(X_filtered_scaled)
        
        # נבדוק אם יש יותר מקלאסטר אחד כדי לחשב את Silhouette Score
        if len(set(labels)) > 1:  # נוודא שיש יותר מקלאסטר אחד (לא רק רעש)
            sil_score = silhouette_score(X_filtered_scaled, labels)
            print(f'eps: {eps_value}, min_samples: {min_samples_value}, Silhouette Score: {sil_score}')
            
            # נשמור את הערכים הטובים ביותר
            if sil_score > best_sil_score:
                best_sil_score = sil_score
                best_eps = eps_value
                best_min_samples = min_samples_value

# תוצאות סופיות
print(f'\nBest eps: {best_eps}, Best min_samples: {best_min_samples}, Best Silhouette Score: {best_sil_score}')


eps: 0.3, min_samples: 4, Silhouette Score: 0.040118796688637166
eps: 0.3, min_samples: 5, Silhouette Score: 0.1408825949879625
eps: 0.3, min_samples: 6, Silhouette Score: 0.158368243393923
eps: 0.4, min_samples: 4, Silhouette Score: 0.25284284044973326
eps: 0.4, min_samples: 5, Silhouette Score: 0.2837533788110633
eps: 0.4, min_samples: 6, Silhouette Score: 0.35605466286836707
eps: 0.5, min_samples: 4, Silhouette Score: 0.3325196444115179
eps: 0.5, min_samples: 5, Silhouette Score: 0.3566897283569976
eps: 0.5, min_samples: 6, Silhouette Score: 0.3566736284559819
eps: 0.6, min_samples: 4, Silhouette Score: 0.33387520909455176
eps: 0.6, min_samples: 5, Silhouette Score: 0.3568447905157725
eps: 0.6, min_samples: 6, Silhouette Score: 0.3568755473227302

Best eps: 0.6, Best min_samples: 6, Best Silhouette Score: 0.3568755473227302


Based on the image you provided, we can observe the results of the DBSCAN algorithm after clustering your shot data, which includes variables like Arc (shot angle) and ShotDistance (distance of the shot), and the assigned clusters (Cluster_Optimized).

Analysis of the Results:
Cluster_Optimized:

At least two primary clusters (0 and 2) were identified based on the similarity of the shot characteristics.
These clusters group together shots with similar features, such as shot angle (Arc) and shot distance (ShotDistance).
Cluster 0:

Characteristics: Medium shot angles (e.g., -101, -68) and relatively longer shot distances (e.g., 113, 115).
The shots in this cluster appear to represent mid-range shots with medium shot angles.
Cluster 2:

Characteristics: More extreme angles (e.g., -163 to -221) and variable shot distances (e.g., 56 to 164).
The shots in this cluster indicate sharper angles and varying shot distances, but they tend to be closer to specific positions on the court.
## Conclusions:
Grouping by Angle and Distance: The algorithm successfully identified two distinct clusters based on the shot angle and distance. Shots in Cluster 0 tend to have less extreme angles and medium to close shot distances, whereas Cluster 2 includes more extreme angles and varying shot distances.

Differences in Shot Style: The differences between the clusters suggest different shooting techniques or positions on the court. Cluster 0 may describe shots that are closer to the basket with more standard angles, while Cluster 2 could represent more challenging, distant shots or shots taken under different circumstances.

Tactical Improvement: If this data reflects game performance, shots in Cluster 2 might be less accurate because of the sharp angles. A deeper analysis could look at the success rates of shots from each cluster and offer insights into the efficiency of these shot types.

Training Opportunities: The shots in Cluster 2 with extreme angles could indicate areas where players may need to work on accuracy from long distances or learn to adjust the balance between angle and shot distance to improve success.

In [2]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import DBSCAN
from sklearn.metrics import silhouette_score

# קריאת הדאטה
data = pd.read_csv(r"C:\Users\guymk\Downloads\data13 (1).csv")  # עדכן את הנתיב בהתאם לצורך

# סינון והכנת הנתונים (התאמת עמודות)
data_filtered = data[['Raw Data', 'Unnamed: 5', 'Unnamed: 28']].copy()
data_filtered.columns = ['HitTime', 'Arc', 'ShotDistance']
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# המרת העמודה HitTime לערכים מספריים
data_filtered['HitTime'] = pd.to_numeric(data_filtered['HitTime'], errors='coerce')
data_filtered = data_filtered.dropna(subset=['HitTime', 'Arc', 'ShotDistance'])

# נרמול הנתונים
scaler = StandardScaler()
X_filtered_scaled = scaler.fit_transform(data_filtered[['HitTime', 'Arc', 'ShotDistance']])

# הרצת DBSCAN עם הערכים שנבחרו
best_eps = 0.5
best_min_samples = 6

dbscan = DBSCAN(eps=best_eps, min_samples=best_min_samples)
labels = dbscan.fit_predict(X_filtered_scaled)

# הוספת התוויות לדאטה המקורי
data_filtered['Cluster'] = labels

# הצגת תוצאות הקלסטרינג
print(data_filtered['Cluster'].value_counts())

# חישוב Silhouette Score עבור התוצאות הטובות ביותר
sil_score = silhouette_score(X_filtered_scaled, labels)
print(f'Silhouette Score for best parameters: {sil_score}')

# שמירת תוצאות הקלסטרינג לתוך קובץ חדש
data_filtered.to_csv('clustered_data.csv', index=False)


Cluster
 0    8085
 1    5550
-1      97
Name: count, dtype: int64
Silhouette Score for best parameters: 0.3566736284559819


## purpose:
We performed a clustering analysis (DBSCAN) on data of shots to the basket with the aim of understanding whether the shots can be divided into different groups based on different characteristics such as the shot time (HitTime), arc (Arc) and distance from the center (Shot distance from center). The goal was to identify groups of shots that can be characterized in a different way and to examine the way in which the shots are distributed in space.

## Clustering results:
Two main clusters: The algorithm identified two main clusters:

Cluster 0: contains 8085 shots.
Cluster 1: contains 5550 shots.
A large amount of injections was divided into two main groups, so that we have a relatively clear division between the injections.

Minimal noise: only 97 shots were classified as noise (Cluster -1). This means that most of the shots are within ordered groups, and it can be seen that they are not defined as isolated points in space, which indicates that the parameters were well chosen to reduce the amount of isolated points.

## Silhouette Score:

The score obtained for the clustering is 0.3567, indicating that the distribution is reasonable but not perfect. The score indicates that the groups are relatively close to each other in space, but there is some overlap between the groups.
It is important to remember that this score indicates a moderate division - the clusters exist, but there may be an overlap in the characteristics of the shots between the different groups.
Interpretation of the findings:
The division into two main clusters can offer interesting insights into the patterns of shots into the basket:

Different clusters of shots: the two groups produced by the DBSCAN may reflect different types of shots in terms of characteristics such as the arc of the shot and distance from the shot position to the center. It is possible that the first group (Cluster 0) reflects shots with a high arc or a specific location, while the second group (Cluster 1) represents shots of a different type.

Low noise: only a small amount of shots were detected as noise, meaning most shots clearly belong to one of the two groups. This result indicates that the data is well organized and there are not many isolated shots that cannot be associated.

## The meaning of the silhouette score:

The score obtained (0.3567) indicates that there is no perfect distribution between the shots, which could imply that there is no clear and sharp difference between the types of shots, or that it is possible that the shots are gradually integrated into the space.
Another possibility is that there is a continuous space where the shots are not unambiguously associated with separate groups.
Conclusions and summary:
Potential for clear division: We were able to divide the shots into two main groups, which indicates that there are certain patterns in the shots that can differentiate between different types of shots.
The little noise indicates good quality: a small amount of shots that failed to belong to any cluster indicates that the data is well organized and insights can be extracted from it.